core/client/src/main/java/alluxio/client/block/BlockStoreContext.java

```diff
@@ -259,6 +269,42 @@ public static void releaseBlockWorkerThriftClient(InetSocketAddress address,
   }
 
   /**
+   * Acquires a block worker thrift client from the block worker thrift client pools dedicated for
+   * heartbeat. If there is no available client instance available in the pool, it tries to create
+   * a new one. And an exception is thrown if it fails to create a new one.
+   *
+   * @param address the address of the block worker
+   * @return the block worker thrift client
+   * @throws IOException if it fails to create a new client instance mostly because it fails to
+   *         connect to remote worker
+   * @throws InterruptedException if this thread is interrupted
+   */
+  public static BlockWorkerClientService.Client acquireBlockWorkerThriftClientHeartbeat(
+      final InetSocketAddress address) throws IOException, InterruptedException {
+    if (!BLOCK_WORKER_THRIFT_CLIENT_HEARTBEAT_POOL.containsKey(address)) {
+      BlockWorkerThriftClientPool pool = new BlockWorkerThriftClientPool(address,
+          Configuration.getInt(PropertyKey.USER_BLOCK_WORKER_CLIENT_POOL_SIZE_MAX) / 2,
```

Comment:
the heartbeat doesn't need that many connections. I think one should be enough after removing the session heartbeat in promote call. But it is not harmful to have a large pool since they will be gc-ed anyway if there is a very high peak usage. I removed this "/2" to avoid confusion for now.